### PR TITLE
Build CNN

### PR DESCRIPTION
**Pull request by parassoni04**
Build a CNN model for MNIST dataset using pyTorch. Solved issue #19

### CODE REVIEW

File: digit_recognition/CNN.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    def forward(self,x):
+        x = self.pool(F.relu(self.conv1(x)))
+        x = self.pool(F.relu(self.conv2(x)))
+
+        x = x.view(x.size(0), -1) # Done for flattening ([batch, 64, 5, 5] -> [batch, 1600])
+
+        x = F.relu(self.fc1(x))
+        x = F.relu(self.fc2(x))
+        x = F.relu(self.fc3(x))
+        x = (self.fc4(x))
```
Comment (DracoRooks) [inline digit_recognition/CNN.py]: shouldn't the output layer be performing softmax operation here?

File: digit_recognition/CNN.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+        return x
+    
+model = NeuralNetwork()
+
+lossFunction = nn.CrossEntropyLoss()
+optimizer = optium.Adam(model.parameters(), lr = 0.001)
```
Comment (DracoRooks) [inline digit_recognition/CNN.py]: use `torch.optim.lr_scheduler` for learning rate decay, preferably `StepLR()` method in it, but that is up to your disgression.

File: digit_recognition/CNN.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+        self.conv1 = nn.Conv2d(1,32,3)
+        self.conv2 = nn.Conv2d(32,64,3)
+
+        self.pool = nn.MaxPool2d(2,2) # Downscale
+
+        self.fc1 = nn.Linear(64*5*5,128)
+        self.fc2 = nn.Linear(128,64)
+        self.fc3 = nn.Linear(64,32)
+        self.fc4 = nn.Linear(32,10)
```
Comment (DracoRooks) [inline digit_recognition/CNN.py]: Please remove all the magic numbers from line 31 to line 39, if you don't know what a "magic number" is, please google it. Please.

File: digit_recognition/CNN.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    for images, labels in trainLoader:
+        optimizer.zero_grad()
+
+        output = model(images)
+        loss = lossFunction(output, labels)
+
+        loss.backward()
+        optimizer.step()
+
+    print(f"Epoch : {e + 1}, Loss : {loss.item()}")
```
Comment (DracoRooks) [inline digit_recognition/CNN.py]: I don't think there's a need to print loss in every epoch, doing once every 10% of epochs should suffice.

File: digit_recognition/CNN.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optium
```
Comment (DracoRooks) [inline digit_recognition/CNN.py]: why are you aliasing `optim` to `optium`...

**Review by DracoRooks — CHANGES_REQUESTED** [review]

File: digit_recognition/model.py [hunk 6]
```diff
@@ -27,7 +29,7 @@ def forward(self, x):
         
         x = x.view(x.size(0), -1)
         
-        x = F.relu(self.fc1(x))
+        x = self.dropout(F.relu(self.fc1(x)))
         x = F.relu(self.fc2(x))
         x = F.relu(self.fc3(x))
```
Comment (DracoRooks) [inline digit_recognition/model.py:34]: You can probably add Dropouts to other linear layers too

**Review by DracoRooks — COMMENTED** [review]
Should this be merged? I think it's good enuf
